tests/infra/partitions.py

```diff
@@ -0,0 +1,196 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import infra.node
+import infra.network
+import iptc
+from dataclasses import field
+from typing import List, Optional
+
+from loguru import logger as LOG
+
+
+CCF_IPTABLES_CHAIN = "CCF-TEST"
+
+CCF_INPUT_RULE = {
+    "protocol": "tcp",
+    "target": CCF_IPTABLES_CHAIN,
+}
+
+# Note: When playing with iptables rules on a remote VM, you may want to:
+#   1. Save the current iptable rules: $ sudo iptables-save > /etc/iptables.conf
+#   2. Setup a cron job to revert the iptables rules regularly, so that you cannot be
+#      logged out of the VM, e.g.:
+#      $ echo "* * * * * root /sbin/iptables-restore /etc/iptables.conf" | sudo tee -a /etc/cron.d/iptables-restore
+# Warning: depending on the cron interval, this may cause partitions test to fail randomly as the iptables rules are
+# deleted under the infra's feet.
+
+
+class Rules:
+    """
+    Set of iptables rules created by the :py:class:`infra.partitions.Partitioner`
+    """
+
+    rules: List[dict] = field(default_factory=list)
+
+    name: Optional[str] = None
+
+    def __init__(self, rules, name=None):
+        self.rules = rules
+        self.name = name
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.drop()
+
+    def drop(self):
+        LOG.info(f'Dropping rules "{self.name or "[unamed]"}"')
+        for rule in self.rules:
+            if iptc.easy.has_rule("filter", CCF_IPTABLES_CHAIN, rule):
+                iptc.easy.delete_rule("filter", CCF_IPTABLES_CHAIN, rule)
+
+
+class Partitioner:
+    """
+    The :py:class:`infra.partitions.Partitioner` provides a convenient way to isolate and
+    create partitions of :py:class:`infra.node.Node` objects. As it relies on iptables,
+    using this class requires admin privileges.
+
+    All member functions return a :py:class:`infra.partitions.Rules` object that can be used
+    in a context manager and which will be automatically dropped when the object goes out of
+    scope.
+
+    Note: It should be managed by a :py:class:`infra.network.Network` instance so that rules
+    outlive nodes to avoid spurious log messages when the network is shutdown.
+    """
+
+    @staticmethod
+    def cleanup():
```

Comment:
The iptables rules are cleaned up by the `infra.network.Network` on shutdown. However, is the test is killed halfway through, the rules won't currently be cleaned up. This is kind of OK if we keep re-running partitions test as the rules are destroyed on creation (see constructor a few lines down) but may affect non-partition tests (in the unlikely but still possible event of port collisions). 

I'd started with adding signal handlers but then remembered that signals are not forwarded by `ctest` to the Python interpreter. A few things we could do:
- Adding a wrapper in `cmake`'s `add_e2e_test` function to save and restore the current iptables rules before and after a partition test. 
- Still add signal handlers so that the rules are probably cleaned up when running without `ctest`. 

Thoughts?

Reply:
Is there any way to tag the rules we set up so we can recognise them somehow?

I would otherwise be in favour of cleaning up any pre-existing rules at test startup, at least for the hosts we know we want to use. Trying to ensure cleanup happens on shutdown every time is difficult to guarantee with absolute certainty.

Reply:
The rules are all under the `CCF-TEST` iptables chain which is now cleaned up on startup. The only other issue is when the privileged `partitions_test` is interrupted and doesn't cleanup the chain, and a _non-privileged_ test runs after it but can't cleanup the left-over chain. 

If this happens on the CI, this would actually be OK as the `partitions_test` runs last. If this happens locally, we may have failing tests but maybe that's not such a big deal as re-running the `partitions_test` would hopefully solve this?

Reply:
Can we add a non-privileged check at the beginning of all tests, that tells us if anything is in that chain and suggests a clean-up command?

Reply:
Unfortunately, even listing the iptables rules requires privileges (unless we want to fiddle with capabilities. which I don't think we want to). See https://unix.stackexchange.com/questions/385109/can-you-list-iptables-as-a-non-root-user-and-why. 

I will list a `iptables` CLI command to cleanup the chain manually in `partition.py` as an alternative, so that it is easy to recover from if this ever happens.

Reply:
I've added a new `cleanup_iptables` Python CLI to cleanup the iptables manually, in case something goes wrong on a local run.